File: laba5.cpp
#include <iostream>
#include <map>
#include <vector>
#include <string>
#include <sstream>

using namespace std;

const int MAX_CELLS = 1920; // Максимальное количество ячеек
const int NUM_ZONES = 2;
const int NUM_SHELVES_PER_ZONE = 6;
const int NUM_SECTIONS_PER_SHELF = 4;
const int NUM_SHELVES_PER_SECTION = 4;

// Структура для хранения информации о товаре в ячейке
struct Item {
    string name;
    int amount;
};

// Класс для управления складом
class Warehouse {
private:
    map<string, vector<Item>> cells; // Контейнер ячеек и содержащихся в них товаров

public:
    void add(string itemName, int amount, const string& address) {
        auto& items = cells[address]; // получаем ссылку на вектор товаров в данной ячейке
        int summ = 0;
        for (const auto& item : items) {
            summ += item.amount;
        }
        // ищем товар с таким же именем
        for (auto& item : items) {
            if (item.name == itemName) {
                // если товар найден, увеличиваем его количество
                if (summ + amount <= 10) {
                    item.amount += amount;
                    return;
                }
                else {
                    cout << "Невозможно добавить товар " << itemName << " в ячейку " << address << ". Превышен лимит в " << 10 << " единиц." << endl;
                    return;
                }
            }
        }
        // если товар не найден, добавляем новый
        if (summ + amount <= 10) {
            items.push_back({itemName, amount});
        }
        else {
            cout << "Ячейка " << address << " заполнена. Невозможно добавить товар " << itemName << "." << endl;
        }
    }

    void remove(string itemName, int amount, const string& address) {
        auto& items = cells[address];
        for (auto it = items.begin(); it != items.end(); it++) {
            if (it->name == itemName) {
                if (it->amount >= amount) {
                    it->amount -= amount;
                    if (it->amount == 0) {
                        items.erase(it);
                    }
                    return;
                }
                else {
                    cout << "В ячейке " << address << " недостаточно товара " << itemName << "." << endl;
                    return;
                }
            }
        }
        cout << "В ячейке " << address << " нет товара " << itemName << "." << endl;
    }

    void info() {
        int totalItems = 0;
        int totalCells = NUM_ZONES * NUM_SHELVES_PER_ZONE * NUM_SECTIONS_PER_SHELF * NUM_SHELVES_PER_SECTION;
        int occupiedCells = 0;

        for (const auto& cell : cells) {
            if (!cell.second.empty()) {
                occupiedCells++;
                for (const auto& item : cell.second) {
                    totalItems += item.amount;
                }
            }
        }

        cout << "Общая загрузка склада: " << (static_cast<double>(totalItems) / (totalCells * 10)) * 100 << "%" << endl;
        cout << "Загрузка зоны A: " << (static_cast<double>(totalItems) / (totalCells * 10)) * 100 << "%" << endl;

        cout << "Содержимое ячеек:" << endl;
        for (const auto& cell : cells) {
            if (!cell.second.empty()) {
                cout << "Ячейка " << cell.first << ": ";
                for (const auto& item : cell.second) {
                    cout << item.name << " (" << item.amount << "), ";
                }
                cout << endl;
            }
        }

        cout << "Пустые ячейки:" << endl;
        for (int i = 0; i < NUM_ZONES; i++) {
            for (int j = 0; j < NUM_SHELVES_PER_ZONE; j++) {
                for (int k = 0; k < NUM_SECTIONS_PER_SHELF; k++) {
                    for (int l = 0; l < NUM_SHELVES_PER_SECTION; l++) {
                        string address = string(1, 'A' + i) + to_string(j + 1) + to_string(k + 1) + to_string(l + 1);
                        if (cells.find(address) == cells.end() || cells.at(address).empty()) {
                            cout << address << ", ";
                        }
                    }
                }
            }
        }
        cout << endl;
    }
};

int main() {
    Warehouse warehouse;
    string input;
    while (true) {
        cout << "Введите команду: ";
        getline(cin, input);
        stringstream ss(input);
        string command;
        ss >> command;
        if (command == "EXIT") {
            break;
        } else if (command == "ADD" || command == "REMOVE") {
            string itemName, address;
            int amount;
            ss >> itemName >> amount >> address;
            if (command == "ADD") {
                warehouse.add(itemName, amount, address);
            } else {
                warehouse.remove(itemName, amount, address);
            }
        } else if (command == "INFO") {
            warehouse.info();
        } else {
            cout << "Неизвестная команда. Попробуйте еще раз." << endl;
        }
    }
    return 0;
}

File: laba5.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
)

const (
	maxCells                = 1920
	numZones                = 2
	numShelvesPerZone       = 6
	numSectionsPerShelf     = 4
	numShelvesPerSection    = 4
)

type Item struct {
	Name   string
	Amount int
}

type Warehouse struct {
	cells map[string][]Item // Контейнер ячеек и содержащихся в них товаров
}

func NewWarehouse() *Warehouse {
	return &Warehouse{cells: make(map[string][]Item)}
}

func (w *Warehouse) Add(itemName string, amount int, address string) {
	items := w.cells[address]
	summ := 0

	for _, item := range items {
		summ += item.Amount
	}

	for i, item := range items {
		if item.Name == itemName {
			if summ+amount <= 10 {
				items[i].Amount += amount
				w.cells[address] = items
				return
			} else {
				fmt.Printf("Невозможно добавить товар %s в ячейку %s. Превышен лимит в %d единиц.\n", itemName, address, 10)
				return
			}
		}
	}

	if summ+amount <= 10 {
		w.cells[address] = append(items, Item{Name: itemName, Amount: amount})
	} else {
		fmt.Printf("Ячейка %s заполнена. Невозможно добавить товар %s.\n", address, itemName)
	}
}

func (w *Warehouse) Remove(itemName string, amount int, address string) {
	items := w.cells[address]

	for i := range items {
		if items[i].Name == itemName {
			if items[i].Amount >= amount {
				items[i].Amount -= amount
				if items[i].Amount == 0 {
					w.cells[address] = append(items[:i], items[i+1:]...) // Удаляем товар из ячейки
				} else {
					w.cells[address] = items // Обновляем ячейку с оставшимися товарами
				}
				return
			} else {
				fmt.Printf("В ячейке %s недостаточно товара %s.\n", address, itemName)
				return
			}
		}
	}
	fmt.Printf("В ячейке %s нет товара %s.\n", address, itemName)
}

func (w *Warehouse) Info() {
	totalItems := 0

	for _, items := range w.cells {
		for _, item := range items {
			totalItems += item.Amount
		}
	}

	totalCells := numZones * numShelvesPerZone * numSectionsPerShelf * numShelvesPerSection

	fmt.Printf("Общая загрузка склада: %.2f%%\n", float64(totalItems)/(float64(totalCells)*10)*100)

	fmt.Println("Содержимое ячеек:")
	for cellAddress, items := range w.cells {
		if len(items) > 0 {
			fmt.Printf("Ячейка %s: ", cellAddress)
			for _, item := range items {
				fmt.Printf("%s (%d), ", item.Name, item.Amount)
			}
			fmt.Println()
		}
	}

	fmt.Print("Пустые ячейки: ")
	for i := 0; i < numZones; i++ {
		for j := 0; j < numShelvesPerZone; j++ {
			for k := 0; k < numSectionsPerShelf; k++ {
				for l := 0; l < numShelvesPerSection; l++ {
					address := fmt.Sprintf("%c%d%d%d", 'A'+i, j+1, k+1, l+1)
					if _, exists := w.cells[address]; !exists || len(w.cells[address]) == 0 {
						fmt.Printf("%s, ", address)
					}
				}
			}
		}
	}
	fmt.Println()
}

func main() {
	var warehouse = NewWarehouse()
	scanner := bufio.NewScanner(os.Stdin)

	for true {
        fmt.Print("Введите команду: ")
        scanner.Scan()
        input := scanner.Text()

        parts := strings.Fields(input)
        if len(parts) == 0 { 
            continue 
        }

        command := parts[0]

        if command == "EXIT" { 
            break 
        } else if command == "ADD" || command == "REMOVE" { 
            if len(parts) != 4 { 
                fmt.Println("Неверный формат команды.") 
                continue 
            }
            itemName := parts[1]
            amount, err := strconv.Atoi(parts[2])
            if err != nil { 
                fmt.Println("Количество должно быть числом.") 
                continue 
            }
            address := parts[3]

            if command == "ADD" { 
                warehouse.Add(itemName, amount, address) 
            } else { 
                warehouse.Remove(itemName, amount, address) 
            }
        } else if command == "INFO" { 
            warehouse.Info() 
        } else { 
            fmt.Println("Неизвестная команда. Попробуйте еще раз.") 
        }
    }
}
